File: internal/git/branch.go
package git

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// GetAllBranches returns a list of all branches (local and remote)
func GetAllBranches(path string) ([]string, error) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return nil, fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	// Get all branches (local and remote)
	cmd := exec.Command("git", "branch", "-a")
	cmd.Dir = workDir

	var stdout bytes.Buffer
	cmd.Stdout = &stdout

	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("failed to get branches: %w", err)
	}

	lines := strings.Split(strings.TrimSpace(stdout.String()), "\n")
	var branches []string

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		// Remove the * prefix for current branch and remote prefixes
		line = strings.TrimPrefix(line, "* ")
		line = strings.TrimPrefix(line, "remotes/")

		// Skip HEAD references
		if strings.Contains(line, "HEAD") {
			continue
		}

		// Remove origin/ prefix for cleaner display
		line = strings.TrimPrefix(line, "origin/")

		// Avoid duplicates
		exists := false
		for _, existing := range branches {
			if existing == line {
				exists = true
				break
			}
		}
		if !exists {
			branches = append(branches, line)
		}
	}

	return branches, nil
}

// GetRemoteBranches returns a list of remote branches
func GetRemoteBranches(path string) ([]string, error) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return nil, fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	// Get remote branches
	cmd := exec.Command("git", "branch", "-r")
	cmd.Dir = workDir

	var stdout bytes.Buffer
	cmd.Stdout = &stdout

	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("failed to get remote branches: %w", err)
	}

	lines := strings.Split(strings.TrimSpace(stdout.String()), "\n")
	var branches []string

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		// Skip HEAD references
		if strings.Contains(line, "HEAD") {
			continue
		}

		// Remove remotes/ prefix for cleaner display
		line = strings.TrimPrefix(line, "remotes/")

		// Avoid duplicates
		exists := false
		for _, existing := range branches {
			if existing == line {
				exists = true
				break
			}
		}
		if !exists {
			branches = append(branches, line)
		}
	}

	return branches, nil
}

// GetBranchDiff returns the diff between two branches
func GetBranchDiff(path, fromBranch, toBranch string) (string, error) {
	// Validate inputs
	if fromBranch == "" {
		return "", fmt.Errorf("source branch name is empty")
	}
	if toBranch == "" {
		return "", fmt.Errorf("target branch name is empty")
	}

	absPath, err := filepath.Abs(path)
	if err != nil {
		return "", fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	// Use a safer approach - compare the branches directly
	cmd := exec.Command("git", "diff", "-U5", "--no-color", fromBranch, toBranch)
	cmd.Dir = workDir

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			if exitErr.ExitCode() == 1 && stdout.Len() > 0 {
				return stdout.String(), nil
			}
		}

		errMsg := strings.TrimSpace(stderr.String())
		if errMsg != "" {
			return "", fmt.Errorf("git diff failed: %s", errMsg)
		}
		return "", fmt.Errorf("git diff failed: %w", err)
	}

	return stdout.String(), nil
}

// GetCurrentBranch returns the current branch name
func GetCurrentBranch(path string) (string, error) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return "", fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	cmd := exec.Command("git", "branch", "--show-current")
	cmd.Dir = workDir

	output, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("failed to get current branch: %w", err)
	}

	return strings.TrimSpace(string(output)), nil
}

File: critica_py/git/branch.py
"""Git branch operations."""

import subprocess
from pathlib import Path
from typing import List


def get_current_branch(path: str = ".") -> str:
    """Get the current branch name.

    Args:
        path: Repository path (default: current directory)

    Returns:
        Current branch name
    """
    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        result = subprocess.run(
            ["git", "branch", "--show-current"],
            cwd=work_dir,
            capture_output=True,
            text=True,
            check=True,
        )
        return result.stdout.strip()
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to get current branch: {e}")


def get_all_branches(path: str = ".") -> List[str]:
    """Get list of all branches (local and remote).

    Args:
        path: Repository path (default: current directory)

    Returns:
        List of branch names
    """
    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        result = subprocess.run(
            ["git", "branch", "-a"],
            cwd=work_dir,
            capture_output=True,
            text=True,
            check=True,
        )

        branches = set()
        for line in result.stdout.strip().split("\n"):
            line = line.strip()
            if not line:
                continue

            # Remove * prefix for current branch
            line = line.removeprefix("* ")
            # Remove remotes/ prefix
            line = line.removeprefix("remotes/")

            # Skip HEAD references
            if "HEAD" in line:
                continue

            # Remove origin/ prefix for cleaner display
            line = line.removeprefix("origin/")

            branches.add(line)

        return sorted(list(branches))
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to get branches: {e}")


def get_remote_branches(path: str = ".") -> List[str]:
    """Get list of remote branches.

    Args:
        path: Repository path (default: current directory)

    Returns:
        List of remote branch names
    """
    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        result = subprocess.run(
            ["git", "branch", "-r"],
            cwd=work_dir,
            capture_output=True,
            text=True,
            check=True,
        )

        branches = set()
        for line in result.stdout.strip().split("\n"):
            line = line.strip()
            if not line:
                continue

            # Skip HEAD references
            if "HEAD" in line:
                continue

            # Remove remotes/ prefix
            line = line.removeprefix("remotes/")

            branches.add(line)

        return sorted(list(branches))
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to get remote branches: {e}")


def get_branch_diff(path: str, from_branch: str, to_branch: str) -> str:
    """Get diff between two branches.

    Args:
        path: Repository path
        from_branch: Source branch name
        to_branch: Target branch name (usually main/master)

    Returns:
        Git diff output as string
    """
    if not from_branch:
        raise ValueError("Source branch name is empty")
    if not to_branch:
        raise ValueError("Target branch name is empty")

    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        # Show changes in from_branch that are not in to_branch
        result = subprocess.run(
            ["git", "diff", "-U5", "--no-color", to_branch, from_branch],
            cwd=work_dir,
            capture_output=True,
            text=True,
            check=False,
        )

        # Exit code 1 with output is normal for git diff
        if result.returncode == 1 and result.stdout:
            return result.stdout

        if result.returncode != 0:
            if result.stderr.strip():
                raise RuntimeError(f"git diff failed: {result.stderr.strip()}")
            raise RuntimeError(f"git diff failed with exit code {result.returncode}")

        return result.stdout
    except FileNotFoundError:
        raise RuntimeError("git command not found. Please install git.")


def get_branch_info(path: str = ".") -> dict:
    """Get information about the current branch and available branches.

    Args:
        path: Repository path (default: current directory)

    Returns:
        Dictionary with branch information
    """
    return {
        "current": get_current_branch(path),
        "all": get_all_branches(path),
        "remote": get_remote_branches(path),
    }
